File: example/g1_low_level/g1_stand_debug.py
#!/home/james/miniconda3/bin/python
"""
G1 Stand Debug - 双进程调试版
增加详细调试输出，定位站立失败原因
"""

import time
import sys
import numpy as np

from unitree_sdk2py.core.channel import ChannelPublisher, ChannelSubscriber
from unitree_sdk2py.core.channel import ChannelFactoryInitialize
from unitree_sdk2py.idl.unitree_hg.msg.dds_ import LowCmd_, LowState_
from unitree_sdk2py.idl.default import unitree_hg_msg_dds__LowCmd_
from unitree_sdk2py.utils.crc import CRC

G1_NUM_MOTOR = 29
dt = 0.002
crc = CRC()

# 站立目标姿态
stand_up_joint_pos = np.array([
    -0.1, 0.0, 0.0, 0.25, -0.15, 0.0,   # left leg
    -0.1, 0.0, 0.0, 0.25, -0.15, 0.0,   # right leg
    0.0, 0.0, 0.0,                       # waist
    0.2, 0.2, 0.0, 0.8, 0.0, 0.0, 0.0,  # left arm
    0.2, -0.2, 0.0, 0.8, 0.0, 0.0, 0.0, # right arm
], dtype=float)

# PD增益
Kp = np.array([
    80, 80, 60, 120, 50, 50,
    80, 80, 60, 120, 50, 50,
    60, 40, 40,
    40, 40, 40, 40, 40, 40, 40,
    40, 40, 40, 40, 40, 40, 40,
], dtype=float)

Kd = np.array([
    2, 2, 2, 3, 1.5, 1.5,
    2, 2, 2, 3, 1.5, 1.5,
    2, 1.5, 1.5,
    1, 1, 1, 1, 1, 1, 1,
    1, 1, 1, 1, 1, 1, 1,
], dtype=float)


class G1StandDebug:
    def __init__(self, network="lo", domain_id=1):
        ChannelFactoryInitialize(domain_id, network)

        self.cmd_pub = ChannelPublisher("rt/lowcmd", LowCmd_)
        self.cmd_pub.Init()
        print("[DEBUG] Publisher initialized on rt/lowcmd")

        self.state_sub = ChannelSubscriber("rt/lowstate", LowState_)
        self.state_sub.Init(self._state_handler, 10)
        print("[DEBUG] Subscriber initialized on rt/lowstate")

        self.current_q = np.zeros(G1_NUM_MOTOR)
        self.current_dq = np.zeros(G1_NUM_MOTOR)
        self.got_state = False
        self.time = 0.0
        self.counter = 0
        self.last_state_time = 0

    def _state_handler(self, msg: LowState_):
        self.last_state_time = time.time()
        for i in range(G1_NUM_MOTOR):
            self.current_q[i] = msg.motor_state[i].q
            self.current_dq[i] = msg.motor_state[i].dq
        self.got_state = True

        self.counter += 1
        if self.counter <= 5 or self.counter % 250 == 0:
            rpy = msg.imu_state.rpy
            quat = msg.imu_state.quaternion
            print(f"[STATE #{self.counter}] IMU rpy: [{rpy[0]:.3f}, {rpy[1]:.3f}, {rpy[2]:.3f}]")
            print(f"  quat: [{quat[0]:.3f}, {quat[1]:.3f}, {quat[2]:.3f}, {quat[3]:.3f}]")
            print(f"  q[0:6] (L-leg):  [{', '.join(f'{x:.3f}' for x in self.current_q[0:6])}]")
            print(f"  q[6:12] (R-leg): [{', '.join(f'{x:.3f}' for x in self.current_q[6:12])}]")
            print(f"  pelvis_height(z): check frame_pos in sensor data")

    def run(self):
        print("[DEBUG] Waiting for motor state on rt/lowstate...")
        timeout_start = time.time()
        while not self.got_state:
            time.sleep(0.01)
            if time.time() - timeout_start > 5.0:
                print("[ERROR] No state received after 5s! DDS communication issue.")
                print("  Check: simulator running? domain_id match? interface match?")
                return

        print(f"[DEBUG] Got first state after {time.time()-timeout_start:.2f}s")
        print(f"[DEBUG] Initial q[0:6]: {self.current_q[0:6]}")
        print(f"[DEBUG] Starting stand-up sequence...")

        duration = 3.0
        log_interval = 100  # 每100步打印一次

        step = 0
        while True:
            step_start = time.perf_counter()
            self.time += dt

            # tanh平滑
            phase = np.tanh(self.time / 1.2)

            cmd = unitree_hg_msg_dds__LowCmd_()
            cmd.mode_pr = 0
            cmd.mode_machine = 5

            for i in range(G1_NUM_MOTOR):
                cmd.motor_cmd[i].mode = 1
                cmd.motor_cmd[i].tau = 0.0
                cmd.motor_cmd[i].dq = 0.0
                cmd.motor_cmd[i].kp = float(Kp[i])
                cmd.motor_cmd[i].kd = float(Kd[i])
                cmd.motor_cmd[i].q = float(
                    phase * stand_up_joint_pos[i] + (1 - phase) * self.current_q[i]
                )

            cmd.crc = crc.Crc(cmd)
            self.cmd_pub.Write(cmd)

            # 调试：打印实际发送的命令
            if step < 3 or step % log_interval == 0:
                q_cmd = [cmd.motor_cmd[i].q for i in range(6)]
                q_actual = list(self.current_q[0:6])
                err = [q_cmd[i] - q_actual[i] for i in range(6)]
                tau_est = [cmd.motor_cmd[i].kp * err[i] for i in range(6)]
                print(f"[CMD step={step} t={self.time:.3f} phase={phase:.3f}]")
                print(f"  q_cmd[0:6]:   [{', '.join(f'{x:.3f}' for x in q_cmd)}]")
                print(f"  q_actual[0:6]:[{', '.join(f'{x:.3f}' for x in q_actual)}]")
                print(f"  error[0:6]:   [{', '.join(f'{x:.3f}' for x in err)}]")
                print(f"  tau_est[0:6]: [{', '.join(f'{x:.1f}' for x in tau_est)}]")
                print(f"  mode_pr={cmd.mode_pr} mode_machine={cmd.mode_machine} crc={cmd.crc}")

            step += 1
            time_until_next = dt - (time.perf_counter() - step_start)
            if time_until_next > 0:
                time.sleep(time_until_next)


if __name__ == "__main__":
    network = "lo"
    domain_id = 1
    if len(sys.argv) >= 2:
        network = sys.argv[1]
        domain_id = 0

    print(f"[G1StandDebug] domain_id={domain_id} interface={network}")
    g1 = G1StandDebug(network, domain_id)
    g1.run()

File: example/g1_low_level/g1_stand_debug.cpp
/*
 * G1 Stand Debug - 双进程调试版
 * 详细打印 DDS 通信、关节状态、命令发送情况
 *
 * Usage: ./g1_stand_debug [network_interface]
 *   无参数: simulation (domain_id=1, interface=lo)
 *   有参数: real robot (domain_id=0, interface=参数值)
 */

#include <cmath>
#include <iostream>
#include <memory>
#include <mutex>
#include <shared_mutex>
#include <string>
#include <chrono>

#include <unitree/robot/channel/channel_publisher.hpp>
#include <unitree/robot/channel/channel_subscriber.hpp>
#include <unitree/idl/hg/LowCmd_.hpp>
#include <unitree/idl/hg/LowState_.hpp>

static const std::string HG_CMD_TOPIC = "rt/lowcmd";
static const std::string HG_STATE_TOPIC = "rt/lowstate";

using namespace unitree::common;
using namespace unitree::robot;
using namespace unitree_hg::msg::dds_;

const int G1_NUM_MOTOR = 29;

template <typename T>
class DataBuffer {
 public:
  void SetData(const T& newData) {
    std::unique_lock<std::shared_mutex> lock(mutex_);
    data_ = std::make_shared<T>(newData);
  }
  std::shared_ptr<const T> GetData() {
    std::shared_lock<std::shared_mutex> lock(mutex_);
    return data_;
  }
 private:
  std::shared_ptr<T> data_;
  std::shared_mutex mutex_;
};

struct MotorState {
  std::array<float, G1_NUM_MOTOR> q = {};
  std::array<float, G1_NUM_MOTOR> dq = {};
};

// 站立目标姿态 (弧度)
std::array<float, G1_NUM_MOTOR> stand_pose = {
    -0.1f, 0.0f, 0.0f, 0.25f, -0.15f, 0.0f,   // left leg
    -0.1f, 0.0f, 0.0f, 0.25f, -0.15f, 0.0f,   // right leg
    0.0f, 0.0f, 0.0f,                          // waist
    0.2f, 0.2f, 0.0f, 0.8f, 0.0f, 0.0f, 0.0f, // left arm
    0.2f, -0.2f, 0.0f, 0.8f, 0.0f, 0.0f, 0.0f // right arm
};

// PD增益 - 为MuJoCo仿真调高
std::array<float, G1_NUM_MOTOR> Kp = {
    80, 80, 60, 120, 50, 50,   // left leg
    80, 80, 60, 120, 50, 50,   // right leg
    60, 40, 40,                // waist
    40, 40, 40, 40, 40, 40, 40, // left arm
    40, 40, 40, 40, 40, 40, 40  // right arm
};

std::array<float, G1_NUM_MOTOR> Kd = {
    2, 2, 2, 3, 1.5f, 1.5f,   // left leg
    2, 2, 2, 3, 1.5f, 1.5f,   // right leg
    2, 1.5f, 1.5f,             // waist
    1, 1, 1, 1, 1, 1, 1,      // left arm
    1, 1, 1, 1, 1, 1, 1       // right arm
};

inline uint32_t Crc32Core(uint32_t* ptr, uint32_t len) {
  uint32_t xbit = 0;
  uint32_t data = 0;
  uint32_t CRC32 = 0xFFFFFFFF;
  const uint32_t dwPolynomial = 0x04c11db7;
  for (uint32_t i = 0; i < len; i++) {
    xbit = 1 << 31;
    data = ptr[i];
    for (uint32_t bits = 0; bits < 32; bits++) {
      if (CRC32 & 0x80000000) {
        CRC32 <<= 1;
        CRC32 ^= dwPolynomial;
      } else {
        CRC32 <<= 1;
      }
      if (data & xbit) CRC32 ^= dwPolynomial;
      xbit >>= 1;
    }
  }
  return CRC32;
}

class G1StandDebug {
 public:
  G1StandDebug(std::string network, int domain_id)
      : time_(0.0), control_dt_(0.002), duration_(3.0),
        counter_(0), got_state_(false) {
    std::cout << "[DEBUG] Initializing DDS with domain_id=" << domain_id
              << " interface=" << network << std::endl;

    ChannelFactory::Instance()->Init(domain_id, network);
    std::cout << "[DEBUG] ChannelFactory initialized" << std::endl;

    lowcmd_pub_.reset(new ChannelPublisher<LowCmd_>(HG_CMD_TOPIC));
    lowcmd_pub_->InitChannel();
    std::cout << "[DEBUG] Publisher on " << HG_CMD_TOPIC << std::endl;

    lowstate_sub_.reset(new ChannelSubscriber<LowState_>(HG_STATE_TOPIC));
    lowstate_sub_->InitChannel(
        std::bind(&G1StandDebug::LowStateHandler, this, std::placeholders::_1), 1);
    std::cout << "[DEBUG] Subscriber on " << HG_STATE_TOPIC << std::endl;

    control_thread_ = CreateRecurrentThreadEx("control", UT_CPU_ID_NONE, 2000,
                                               &G1StandDebug::Control, this);
    cmd_thread_ = CreateRecurrentThreadEx("cmd_writer", UT_CPU_ID_NONE, 2000,
                                           &G1StandDebug::WriteCommand, this);
    std::cout << "[DEBUG] Control threads started (2ms period)" << std::endl;
    std::cout << "[DEBUG] Waiting for state on " << HG_STATE_TOPIC << "..." << std::endl;
  }

 private:
  void LowStateHandler(const void* message) {
    LowState_ state = *(const LowState_*)message;

    MotorState ms;
    for (int i = 0; i < G1_NUM_MOTOR; ++i) {
      ms.q[i] = state.motor_state()[i].q();
      ms.dq[i] = state.motor_state()[i].dq();
    }
    motor_state_.SetData(ms);

    if (!got_state_) {
      got_state_ = true;
      std::cout << "[DEBUG] *** FIRST STATE RECEIVED ***" << std::endl;
      std::cout << "[DEBUG] mode_machine=" << (int)state.mode_machine() << std::endl;
      std::cout << "[DEBUG] Initial q[0:6]: ";
      for (int i = 0; i < 6; i++) std::cout << ms.q[i] << " ";
      std::cout << std::endl;
      std::cout << "[DEBUG] Initial q[6:12]: ";
      for (int i = 6; i < 12; i++) std::cout << ms.q[i] << " ";
      std::cout << std::endl;
    }

    if (++counter_ % 500 == 0) {
      auto& rpy = state.imu_state().rpy();
      printf("[STATE #%d] IMU rpy: %.3f %.3f %.3f | q[knee_L]=%.3f q[knee_R]=%.3f\n",
             counter_, rpy[0], rpy[1], rpy[2], ms.q[3], ms.q[9]);
    }
  }

  void Control() {
    auto ms = motor_state_.GetData();
    if (!ms) return;

    time_ += control_dt_;
    float phase = std::tanh(time_ / 1.2f);

    LowCmd_ cmd;
    cmd.mode_pr() = 0;  // PR mode
    cmd.mode_machine() = 5;  // G1 29-DOF

    for (int i = 0; i < G1_NUM_MOTOR; ++i) {
      cmd.motor_cmd()[i].mode() = 1;
      cmd.motor_cmd()[i].tau() = 0.0f;
      cmd.motor_cmd()[i].dq() = 0.0f;
      cmd.motor_cmd()[i].kp() = Kp[i];
      cmd.motor_cmd()[i].kd() = Kd[i];
      cmd.motor_cmd()[i].q() = phase * stand_pose[i] + (1.0f - phase) * ms->q[i];
    }

    cmd.crc() = Crc32Core((uint32_t*)&cmd, (sizeof(cmd) >> 2) - 1);
    motor_cmd_.SetData(cmd);

    // 调试输出
    static int step = 0;
    if (step < 5 || step % 500 == 0) {
      printf("[CMD step=%d t=%.3f phase=%.3f] mode_pr=%d mode_machine=%d\n",
             step, time_, phase, (int)cmd.mode_pr(), (int)cmd.mode_machine());
      printf("  q_cmd[0:6]:   ");
      for (int i = 0; i < 6; i++) printf("%.3f ", cmd.motor_cmd()[i].q());
      printf("\n  q_actual[0:6]:");
      for (int i = 0; i < 6; i++) printf("%.3f ", ms->q[i]);
      printf("\n  error[0:6]:   ");
      for (int i = 0; i < 6; i++) printf("%.3f ", cmd.motor_cmd()[i].q() - ms->q[i]);
      printf("\n  kp[0:6]:      ");
      for (int i = 0; i < 6; i++) printf("%.0f ", cmd.motor_cmd()[i].kp());
      printf("\n");
    }
    step++;
  }

  void WriteCommand() {
    auto mc = motor_cmd_.GetData();
    if (mc) {
      lowcmd_pub_->Write(*mc);
    }
  }

  double time_, control_dt_, duration_;
  int counter_;
  bool got_state_;
  DataBuffer<MotorState> motor_state_;
  DataBuffer<LowCmd_> motor_cmd_;
  ChannelPublisherPtr<LowCmd_> lowcmd_pub_;
  ChannelSubscriberPtr<LowState_> lowstate_sub_;
  ThreadPtr control_thread_, cmd_thread_;
};

int main(int argc, char const* argv[]) {
  std::string network = "lo";
  int domain_id = 1;

  if (argc >= 2) {
    network = argv[1];
    domain_id = 0;
  }

  std::cout << "[G1StandDebug] domain_id=" << domain_id
            << " interface=" << network << std::endl;

  G1StandDebug g1(network, domain_id);

  // 等待10秒收集调试信息
  for (int i = 0; i < 10; i++) {
    sleep(1);
    std::cout << "[DEBUG] Running... " << (i+1) << "s" << std::endl;
  }

  std::cout << "[DEBUG] Done." << std::endl;
  return 0;
}
